Provide rpm build target

glimchb opened the pull request:
Makes creating an RPM package easy.

glimchb committed:
Provide rpm build target

Makes creating an RPM package easy.

lsgunth commented:
Thanks, this is great. I just have one concern:

Is it absolutely necessary to mangle the version number replacing dashes with dots? I really don't like how the version number goes from

`0.9.5-rc1-13-g99c0c4d831e9`

to 

`0.9.5.rc1.14.g364d08f88dbb`

Using a dash to separate the 'rc' from the proper version number is pretty common practice and is a fair bit easier to read.

lsgunth commented:
After a small amount of research, I think what we should do is have the script replace the dashes with a tilde (~). But I'd like to do this only for packaging not generally. The VERSION-GEN script can create a PKG_VERSION with this replacement made for use in the spec file and maybe some future debian package generator.

Thoughts?

lsgunth commented:
Ugh, no, Debian uses the tilde in a nice way but RPM seems to say it's not allowed. According to [this](https://fedoraproject.org/wiki/Packaging:Versioning) we need to strip off everything after the first dash and put it in the "extraver" section. So this will have to be done in the VERSION-GEN script.

Sigh, when did versioning become so hard.

glimchb commented:
Yeah, I know...
RPM have more strict version rules...
That's why I had to replace with dots...
I was looking as a reference to this nvme-cli project and that's what they choose to do (dots): https://github.com/linux-nvme/nvme-cli/blob/master/nvme.spec.in

Another thing we can do is as you suggested to use "extraver"...

I'm open for suggestions... And thanks a lot for your time reviewing my changes.

glimchb commented on switchtec.spec.in:
```diff
@@ -0,0 +1,41 @@
+%install
+rm -rf $RPM_BUILD_ROOT
+make install DESTDIR=%{buildroot}
+ln -sfn libswitchtec.so.* %{buildroot}/usr/local/lib/libswitchtec.so
```
"ln -sfn" is a workaround... need to find a better way...

lsgunth replied:
Should you not just copy it? Perhaps with --preserve-links?

lsgunth commented:
Yeah first, as I said, I think this needs to be RPM specific and not mangle the version universally.

Then you'll want to split the string by the first dash and replace dashes with dots in the second half.

Use the first half as the version.

Then if there is no second half, make RELEASE=1;  otherwise RELEASE=0.{rcstring}

I think that's what RPM wants you to do. I found a big discussion on adding the tilde usage (which Debian has had for 15 years) that describes some of this.

Logan

glimchb commented:
fixed it. now RPM have VERSION and RELEASE

lsgunth commented:
Ok, I have a couple things:

1) I think you have to prefix RELEASE with '0.' and set it to 1 if there is no release. Otherwise v0.9-rc5 will be assumed to be newor than v0.9 which we don't want

2) I noticed we needed to set 'dist' as a phony target. This is my bug from a while ago so I've pushed a commit to this repo to fix it.

3) When I attempt 'make rpm' I get an error saying ldconfig doesn't exist. I think when packaging we do not want to run ldconfig. So we probably need an install target that doesn't do ldconfig for packaging.

4) We still need to sort out this TBD 'ln' command. I'm not really sure what that is doing.

Thanks,

Logan

glimchb commented:
Fixed all but 4

when I remove the hack I get this error I'm tryng to find how to solve:
```bash
RPM build errors:
    Symlink points to BuildRoot: /usr/local/lib/libswitchtec.so -> /home/borisg/rpmbuild/BUILDROOT/switchtec-0.9.5-0.rc1.14.gb148a3f.dirty.el7.centos.x86_64/usr/local/lib/libswitchtec.so.0.9.5
make: *** [rpm] Error 1
```

glimchb commented:
all fixed

lsgunth commented:
I don't see how you addressed item 3 for the ldconfig issue and you've committed some unnecessary merge commit you'll want to get rid of...

I think the reason you get that 'ln' error is because the makefile creates symlinks with full paths. This is fine for just regular installs but we should probably use relative paths if we want to work with rpm.

Try this patch and see if it will work without the extra ln line in the spec file.

```
diff --git a/Makefile.in b/Makefile.in
index 1c666ed01992..60ae8be0ddd3 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -151,8 +151,7 @@ install-bin: compile
        @$(NQ) echo "  INSTALL  $(LIBDIR)/$(IMPLIBNAME).$(VERSION)"
        $(Q)install $(IMPLIBNAME) $(LIBDIR)/$(IMPLIBNAME).$(VERSION)
        @$(NQ) echo "  INSTALL  $(LIBDIR)/$(IMPLIBNAME)"
-       $(Q)ln -fs $(LIBDIR)/$(IMPLIBNAME).$(VERSION) \
-           $(LIBDIR)/$(IMPLIBNAME)
+       $(Q)ln -fs $(IMPLIBNAME).$(VERSION) $(LIBDIR)/$(IMPLIBNAME)
        @$(NQ) echo "  INSTALL  $(INCDIR)/switchtec"
        @$(Q)mkdir -p $(INCDIR)
        @$(Q)cp -r inc/switchtec $(INCDIR)

```

glimchb commented:
1. I removed the merge commit. 
2. issue 3 was solved easy by adding "LDCONFIG=" in the spec file 
```spec
%makeinstall DESTDIR=%{buildroot} LDCONFIG=
```
3. I fixed link issue as well in spec file according to info i found in google. I tried the same fix as you proposed but it doesn't work

lsgunth commented:
The ldconfig issue is not fixed. I still see the same issue when doing 'make rpm'.

I think you should try the potential solution to the link issue I posted as I think it's a much better solution. That extra ln line is still ugly and confusing.

lsgunth commented:
Can you also rebase onto the master in this repo? So there isn't a conflict and travis will run again.

glimchb commented:
1. The ldconfig issue should be fixed. I tested on several machines. Try to clean your repo , rerun ./configure and re-run make rpm again please.
2. I agree that extra line is ugly but the patch suggested is not working. I tried to fix it for a few hours without success. I suggest we merge it like this
3. I rebased

lsgunth commented:
Ok, the ldconfig thing does work now.

I also dug into the ln issue and found my patch does work. However it was a pain because rpmbuild seems to cache something in some strange place that it wouldn't see changes to the Makefile unless the version number changed. Even if I nuked the rpmbuild directory it creates in my home directory. Nasty.

Anyway, after sorting that out I reproduced the error, and then fixed it with my patch. I've pushed branch 'rpm' to this repo, can you please test it and confirm it works for you? Then I'll merge it.

Thanks,

Logan

glimchb commented:
Hi,

Unfortunately you patch doesn't work.
the rpm build is a success but rpm can't be installed.

during the rpmbuild you can see this problematic line (look for libswitchtec.so here):

```bash
Requires: libc.so.6()(64bit) libc.so.6(GLIBC_2.14)(64bit) libc.so.6(GLIBC_2.2.5)(64bit) libc.so.6(GLIBC_2.3)(64bit) libc.so.6(GLIBC_2.3.4)(64bit) libc.so.6(GLIBC_2.4)(64bit) libc.so.6(GLIBC_2.7)(64bit) libswitchtec.so()(64bit) rtld(GNU_HASH)
````

So during installation this dependency is wrong and can't be found.

```bash
yum install  ~/rpmbuild/RPMS/x86_64/switchtec-0.8-0.234.gf35690f.el7.centos.x86_64.rpm
...
Resolving Dependencies
--> Running transaction check
---> Package switchtec.x86_64 0:0.8-0.234.gf35690f.el7.centos will be installed
--> Processing Dependency: libswitchtec.so()(64bit) for package: switchtec-0.8-0.234.gf35690f.el7.centos.x86_64
...
--> Finished Dependency Resolution
Error: Package: switchtec-0.8-0.234.gf35690f.el7.centos.x86_64 (/switchtec-0.8-0.234.gf35690f.el7.centos.x86_64)
           Requires: libswitchtec.so()(64bit)
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
```


This is the exact thing I was trying to avoid...

lsgunth commented:
Sorry, but I'm not merging this until I'm convinced we have the correct solution. I'm fairly certain at this point that you're 'ln' line is not correct.

So I don't understand why your rpm has that extra dependency. Nor do I undrestand how the differences could possibly cause that. My output RPM does not have that dependency:

```
$ rpm -qp --requires switchtec-0.9.5-0.rc1.15.gf35690f45469.x86_64.rpm 
libc.so.6()(64bit)
libc.so.6(GLIBC_2.14)(64bit)
libc.so.6(GLIBC_2.2.5)(64bit)
libc.so.6(GLIBC_2.3)(64bit)
libc.so.6(GLIBC_2.7)(64bit)
libncurses.so.5()(64bit)
libncurses.so.5(NCURSES_5.0.19991023)(64bit)
libtinfo.so.5()(64bit)
libtinfo.so.5(NCURSES_TINFO_5.0.19991023)(64bit)
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(PayloadFilesHavePrefix) <= 4.0-1
rtld(GNU_HASH)

$ rpm -qp --provides switchtec-0.9.5-0.rc1.15.gf35690f45469.x86_64.rpm 
libswitchtec.so.0.9.5()(64bit)
switchtec
switchtec = 0.9.5-0.rc1.15.gf35690f45469
switchtec(x86-64) = 0.9.5-0.rc1.15.gf35690f45469

$ rpm -qilp switchtec-0.9.5-0.rc1.15.gf35690f45469.x86_64.rpm
Name        : switchtec
Version     : 0.9.5
Release     : 0.rc1.15.gf35690f45469
Architecture: x86_64
Install Date: (not installed)
Group       : Development/Tools
Size        : 402035
License     : MIT
Signature   : (none)
Source RPM  : switchtec-0.9.5-0.rc1.15.gf35690f45469.src.rpm
Build Date  : Wed 18 Apr 2018 09:59:06 AM MDT
Build Host  : cgy1-donard.priv.deltatee.com
Relocations : (not relocatable)
URL         : https://github.com/Microsemi/switchtec-user
Summary     : Userspace code for the Microsemi PCIe switch
Description :
The code hosted here provides an easy to use CLI and C library for
communicating with Microsemi's Switchtec management interface.
/etc/bash_completion.d/switchtec
/usr/local/bin/switchtec
/usr/local/include/switchtec/arbitration.h
/usr/local/include/switchtec/bind.h
/usr/local/include/switchtec/endian.h
/usr/local/include/switchtec/errors.h
/usr/local/include/switchtec/gas.h
/usr/local/include/switchtec/log.h
/usr/local/include/switchtec/mrpc.h
/usr/local/include/switchtec/pmon.h
/usr/local/include/switchtec/portable.h
/usr/local/include/switchtec/registers.h
/usr/local/include/switchtec/switchtec.h
/usr/local/include/switchtec/utils.h
/usr/local/lib/libswitchtec.a
/usr/local/lib/libswitchtec.so
/usr/local/lib/libswitchtec.so.0.9.5

```

It's especially strange given that the 'switchtec' binary in this package is statically linked and nothing actually depends on the shared library inside the package.  So I can't understand why the auto dependency solver would pick it up for you.

Can you send the output of the same commands as well as the complete output of 'make rpm'?

Logan

lsgunth commented:
I'm not sure if this is the best solution either but we could try adding 

`AutoReqProv: no `

To the spec file...

lsgunth commented:
I made a bunch of updates to the 'rpm' branch. Please try it again with that too.

Thanks,

Logan

glimchb commented:
I was OOO, will test this today

glimchb commented:
I tested https://github.com/Microsemi/switchtec-user/commits/rpm and it works!!

Thanks for your patience and help!!

lsgunth commented:
Thanks for your patience getting it correct and your work adding new features! I've merged it into devel and master.
